pkg/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,86 @@
+package ratelimit
+
+import (
+	"fmt"
+
+	sensor "github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1"
+)
+
+const (
+	secondsInMinute float64 = 60
+	secondsInHour   float64 = 3600
+)
+
+type RateLimitCalculator struct {
+	defaultRateLimit sensor.RateLimit
+}
+
+func NewRateLimitCalculatorOrDie(defaultUnit string, defaultLimitValue int32) *RateLimitCalculator {
+	if !validRateLimitUnit(defaultUnit) {
+		panic(fmt.Errorf("invalid defaultUnit: %s", defaultUnit))
+	}
+
+	return &RateLimitCalculator{
+		defaultRateLimit: sensor.RateLimit{
+			Unit:            sensor.RateLimiteUnit(defaultUnit),
+			RequestsPerUnit: defaultLimitValue,
+		},
+	}
+}
+
+// Calculates the RateLimit based on min(sensorValue, maxRateLimit) where
+// maxRateLimit is the namespaceValue if set, otherwise defaultRateLimit.
+func (r *RateLimitCalculator) Calculate(namespaceValue, sensorValue *sensor.RateLimit) (sensor.RateLimit, error) {
+	maxRateLimit := r.defaultRateLimit
+	if namespaceValue != nil {
+		// Namespace value overrides the default
+		maxRateLimit = *namespaceValue
+	}
+
+	if sensorValue == nil {
+		return maxRateLimit, nil
+	}
+
+	return min(maxRateLimit, *sensorValue)
+}
+
+func validRateLimitUnit(unit string) bool {
+	rateLimitUnit := sensor.RateLimiteUnit(unit)
+
+	switch rateLimitUnit {
+	case sensor.Second, sensor.Minute, sensor.Hour:
+		return true
+	default:
+		return false
+	}
+}
+
+func min(a, b sensor.RateLimit) (sensor.RateLimit, error) {
+	aRequestsPerSecond, err := calculateRequestsPerSecond(a)
+	if err != nil {
+		return sensor.RateLimit{}, err
+	}
+
+	bRequestsPerSecond, err := calculateRequestsPerSecond(b)
+	if err != nil {
+		return sensor.RateLimit{}, err
+	}
+
+	if aRequestsPerSecond < bRequestsPerSecond {
+		return a, nil
+	}
+	return b, nil
+}
+
+func calculateRequestsPerSecond(input sensor.RateLimit) (float64, error) {
+	switch input.Unit {
+	case sensor.Second:
+		return float64(input.RequestsPerUnit), nil
+	case sensor.Minute:
+		return float64(input.RequestsPerUnit) / secondsInMinute, nil
+	case sensor.Hour:
+		return float64(input.RequestsPerUnit) / secondsInHour, nil
+	default:
+		return 0.0, fmt.Errorf("invalid RateLimit unit: %s", input.Unit)
+	}
```

Comment:
This logic can probably be simplified.  There are only three valid unit types, Second, Minute, and Hour.    If the `input` has an empty unit, we should just pick a default and avoid returning an error.

It may also be worth testing the admission of a Sensor with an invalid unit.  This may be caught by a validating webhook.

Reply:
Changed to default to Second.

Removed the error return from here and all other methods. Updated unit tests.